Fix #2940: Fix chapter lesson blurring for SVG thumbnails

BenHenning opened the pull request:
Fix #2940

Adds support for blurring SVG images. The main issue is that our previous transformation is a ``BitmapTransformation``, and those don't apply to SVGs. This actually took me some time to figure out (https://github.com/bumptech/glide/issues/2631#issuecomment-803228591 was one of my braindumps). Glide does have support for attaching custom transformations to custom types (such as our ``ScalableVectorGraphic`` class). Beyond that, I couldn't find a good way to blur the SVG, so I added a pathway to prerender the SVG to a bitmap and then blur that. Not only does this consolidate our blurring to one routine (which adds consistency), it also is likely more performant to render bitmaps than SVGs. I might add bitmap pre-rendering for all SVGs since sometimes the picture rendering fails, and bitmap rendering seems more reliable.

Testing is tricky here. We ideally need Glide testing to verify upstream that the drawable is created & arranged correctly (tracked by #1523), and we need screenshot testing to verify visual correctness especially for blurring (#1815). I'm not clear on what else I can test--if anyone has ideas, please let me know.

In order to test this locally, you'll need to download the alpha assets locally. This is a bit tricky to do, so please follow up with me if you want to verify locally. Otherwise, you can verify this is correct via the alpha test binary after all of the alpha PRs are merged in.

One side change: I updated the bitmap transformation to correctly use a message digest/equals/hashCode as recommended by Glide documentation: https://bumptech.github.io/glide/doc/transformations.html#required-methods. From my understanding, the previous implementation is unlikely to run into any issues but the changes make the transformation more correct per documentation.

Screenshot of blurring with SVG:

![image](https://user-images.githubusercontent.com/12983742/114964785-41824600-9e24-11eb-9a16-c0c143f5dc6c.png)

BenHenning committed:
Add support for loading proto versions of lessons.

BenHenning committed:
Upgrade Espresso versions to work around build issues.

Note that I can't actually verify Espresso tests are working with these
versions since the Bazel project doesn't yet have Espresso tests set up.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Lint fix post-copying from #2927.

BenHenning committed:
Fix Glide in Bazel.

BenHenning committed:
Fix concept card name escaping.

Also, refactor HtmlParser to better support/handling different tag types
& prepare for LaTeX SVG image support.

BenHenning committed:
Copy over test fix from #2927.

BenHenning committed:
Add support for rendering LaTeX via SVGs.

Much of this is copied from #2927 so see that PR's commit history & the
PR corresponding to this commit for specific details.

BenHenning committed:
Ensure all thumbnails are opaque.

BenHenning committed:
Fix blurring for SVGs.

SVGs did not previous support blurring, & this change introduces a way
to leverage the existing bitmap blurrer to introduce blurring
functionality in a way that's consistent with bitmap thumbnails.

BenHenning committed:
Lint fixes post copying from #2927.

BenHenning committed:
Remove image loading annotation.

This will be added in a later branch in the chain.

BenHenning committed:
Add caching module to needed test suites.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Re-add image load annotation for this PR.

BenHenning committed:
Merge branch 'support-latex-svg-rendering' into fix-thumbnail-alpha-overrides

BenHenning committed:
Merge branch 'fix-thumbnail-alpha-overrides' into fix-locked-lesson-blurring

BenHenning commented:
/cc @rt4914

rt4914 commented:
@BenHenning For this can you mention the steps to load the alpha content so that I can check on that.

BenHenning committed:
Add support for importing text protos as assets.

This actually provides Bazel-only support for converting text protos to
binary and including those as assets so that we can avoid checking in
binary protos to the codebase.

BenHenning committed:
Add gitignore for Android Studio Bazel plugin output.

BenHenning committed:
Import textproto versions of existing dev assets.

This imports local conversions of all existing dev assets as text proto
leveraging the Bazel-only conversion system to convert these to binary
protos and include them as assets.

This is a necessary step to both eventually replace JSON assets with
text proto, and to test that the new proto loading pipeline is working
as designed.

BenHenning committed:
Undo Espresso version change since it breaks Bazel tests.

BenHenning committed:
Fix tests.

Various attempts to fix existing tests + move some of the domain tests
over to actually using proto when running with Bazel.

App module tests are hanging currently for unknown reasons (only in
Bazel).

BenHenning committed:
Use correct Espresso core.

3.1.0 causes tests to hang indefinitely when run with Bazel.

BenHenning commented:
A bug I found while testing with production alpha: sometimes images won't load in the list (& will either white out or just be the filled color). This seems to only affect the story list, and sometimes they appear after scrolling & waiting and other times they don't (even with app restart). It seems like something is failing when converting to bitmap & it never recovers.

rt4914 commented:
@BenHenning Does this PR still needs a review?

BenHenning committed:
Move coroutine dispatchers.

This moves the coroutine dispatchers to a new threading subpackage. It
also simplifies the TestCoroutineDispatcher interface & implementations
(which required moving runUntilIdle to CoroutineExecutorServiceTest).

The test coroutine annotations were also moved to their own files.

BenHenning committed:
Refactor testing utilities build graph.

This introduces new Bazel libraries for:
- Test dispatchers
- Fake system/Oppia clocks
- Robolectric dependencies

BenHenning committed:
Introduce dedicated test for TestCoroutineDispatcher.

This also fixes some threading issues in the dispatcher, and clarifies
some of its API that was previously unclear (and can lead to subtle race
conditions).

BenHenning committed:
Add tests for TestCoroutineDispatcher.

This required a bunch of refactoring, adding the first examples of
dedicated Bazel test build files, and introducing a new test pattern for
sharing code between different implementations. Further, it refined the
API for TesCoroutineDispatcher & fixed some issues in the
implementations (especially Robolectric).

Tests verified as non-flaky in Robolectric across 1000 runs, and 0.2%
flaky across 1000 runs for Espresso (though that may be fixed; waiting
on a follow-up run to confirm stability).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Workaround ktlint semicolon issue.

BenHenning committed:
Lint fixes.

BenHenning committed:
Fix test post-merge.

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Move CoroutineDispatcher to be part of threading.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into fix-protobuf-duplication-issue

BenHenning committed:
Undo version upgrade.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	.gitignore
	app/src/sharedTest/java/org/oppia/android/app/story/StoryActivityTest.kt

BenHenning committed:
Fix broken tests.

BenHenning committed:
Lint fixes.

BenHenning committed:
Add a bit more proto loading test coverage.

BenHenning committed:
Comment fix.

BenHenning committed:
Lint fixes.

BenHenning committed:
Address earlier TODO.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	tools/BUILD.bazel

BenHenning committed:
Post-merge restructure.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	data/BUILD.bazel

BenHenning committed:
Post-merge fixes.

This removes unneeded dependencies from PersistentCacheStore, including
Glide.

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Add tests for new tag handlers.

Also, add tests for HtmlParser, and clean up some existing tests.

BenHenning committed:
Fix broken test build.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into fix-glide-image-loading

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Post-merge lint fixes.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

Conflicts:
	app/BUILD.bazel
	app/src/sharedTest/java/org/oppia/android/app/parser/HtmlParserTest.kt
	testing/BUILD.bazel
	utility/BUILD.bazel

BenHenning committed:
Lint fixes.

BenHenning committed:
Address reviewer comment.

BenHenning committed:
Post-merge fix.

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

Conflicts:
	data/BUILD.bazel
	utility/src/main/java/org/oppia/android/util/caching/CacheAssetsLocally.kt

BenHenning committed:
Address pre-merge TODO.

BenHenning committed:
Disable local image loading.

Also, add specific LaTeX line in the first fractions interaction so that
the new SVG support can be tested.

BenHenning committed:
Fix typo.

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Docs + tests.

This adds more documentation & context for code added before without it.
It also adds a bunch of new testing coverage for changes. The tests are
particularly around the html part of the LaTeX loading pipeline, but not
the image parts.

More of both will be needed specifically for the image loading pipeline
changes.

BenHenning committed:
Docs + tests + refactor.

This adds documentation for all remaining components introduced in #2981,
adds some testing for UrlImageParser, and cleans up the image loading
pipeline to be a bit simpler & cleaner (though this will probably break
down-stream PRs relying on image transformations).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Lint fixes.

BenHenning committed:
Fix lint failure & re-enable assets.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Merge branch 'support-latex-svg-rendering' into fix-thumbnail-alpha-overrides

BenHenning committed:
Clean up + tests.

Add tests where reasonable given current testing limitations and
developing time availability, and remove some dead code. Add new testing
utility (GenericViewMatchers) + corresponding Bazel package.

BenHenning committed:
Consolidate testing module view matchers.

This moves two other view matchers that are part of the testing module
into the new Espresso package.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'fix-thumbnail-alpha-overrides' into fix-locked-lesson-blurring

Conflicts:
	utility/src/main/java/org/oppia/android/util/parser/SvgPictureDrawable.kt

BenHenning committed:
Post-merge fixes + docs.

Add TODOs for tests that can't easily be added. Fill in missing
documentation. Fix post-merge issues to ensure SVGs can actually be
blurred (will be verified after this commit).

BenHenning commented:
Yup @rt4914, but it looks like no codeowners reviews are necessary. I'll ask @viktoriias to review since she introduced the original blurring code, and you since you're reviewing #2981 and that's introducing most of the SVG stuff used here.

BenHenning commented:
@rt4914 & @viktoriias PTAL.

rt4914 commented on utility/src/main/java/org/oppia/android/util/parser/BitmapBlurrer.kt:
```diff
@@ -6,52 +6,48 @@ import android.renderscript.Allocation
+private const val BLUR_RADIUS = 20f
 
+/** Utility used to blur [Bitmap]s. */
+internal class BitmapBlurrer(private val context: Context) {
```
Though this class name makes sense to me but reading it sounds a bit awkward.

BenHenning replied:
Might be because 'blurrer' isn't a word. :) Any suggestions for alternative names?

rt4914 replied:
`BitmapBlurEffect` , `BitmapBlurImpl` ?

BenHenning replied:
'Impl' doesn't really fit since it's not implementing an interface/API. 'BlurEffect' is a bit off, too, since 'effect' actually means something in Android and this isn't that (https://developer.android.com/reference/android/media/effect/Effect).

I think I'll keep this now (for time), but if you have additional thoughts I'm happy to discuss.

rt4914 reviewed (approved):
LGTM.

viktoriias commented:
LGTM

BenHenning committed:
Merge branch 'develop' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning commented:
Looks like the Gradle failure is due to the uncommon LogUploadWorkerTest flake being tracked by #2032.

BenHenning committed:
Merge branch 'develop' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Merge branch 'develop' into support-latex-svg-rendering

BenHenning committed:
Merge branch 'support-latex-svg-rendering' into fix-thumbnail-alpha-overrides

BenHenning committed:
Merge branch 'develop' into fix-thumbnail-alpha-overrides

BenHenning committed:
Merge branch 'fix-thumbnail-alpha-overrides' into fix-locked-lesson-blurring

BenHenning commented:
Thanks for the reviews!

See https://github.com/oppia/oppia-android/pull/2980#issuecomment-825478246 for context on why I'm both enabling auto-merge and cancelling the Bazel test CI run.

Note also that the code reviewer request above isn't right--it was a temporary request due to the base branch changing without the branch itself being updated to be properly based on develop post the squash merge of #2983.

BenHenning commented:
Hmm. CI isn't triggering for some reason. Closing & reopening PR to see if that triggers it.

BenHenning commented:
Note that I do plan on letting the full CI checks run on develop now that this & the previous 4 PRs have been merged.
